server/src/metrics/edge_metrics.rs

```diff
@@ -66,6 +65,234 @@ impl Clone for RequestStats {
     }
 }
 
+#[derive(Debug, Default, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct DataPoint {
+    pub interval: [u64; 2],
+    pub requests: AtomicU64,
+}
+
+impl Clone for DataPoint {
+    fn clone(&self) -> Self {
+        Self {
+            interval: self.interval,
+            requests: AtomicU64::new(self.requests.load(Ordering::Relaxed)),
+        }
+    }
+}
+
+#[derive(Debug, Default, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ConsumptionGroup {
+    pub metered_group: String,
+    pub data_points: Vec<DataPoint>,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, Hash, Eq, PartialEq)]
+pub enum ConnectionMetricsType {
+    Features,
+    Metrics,
+}
+
+impl ConnectionMetricsType {
+    fn from_endpoint(endpoint: &str) -> Option<Self> {
+        if endpoint.contains("/features") || endpoint.contains("/delta") {
+            Some(Self::Features)
+        } else if endpoint.contains("/metrics") {
+            Some(Self::Metrics)
+        } else {
+            None
+        }
+    }
+}
+
+const DEFAULT_METRICS_INTERVAL: u64 = 60000;
+const DEFAULT_FEATURES_INTERVAL: u64 = 15000;
+const BUCKET_SIZE_METRICS: u64 = 60000;
+const BUCKET_SIZE_FEATURES: u64 = 5000;
+const MAX_BUCKET_INTERVAL: u64 = 3600000;
+
+#[derive(Debug, Clone, Copy, Hash, Eq, PartialEq, Deserialize, Serialize)]
+pub struct BucketRange {
+    start: u64,
+    end: u64,
+}
+
+impl BucketRange {
+    pub fn new(start: u64, end: u64) -> Self {
+        assert!(
+            start <= end,
+            "Bucket start must be less than or equal to end"
+        );
+        Self { start, end }
+    }
+
+    pub fn as_array(&self) -> [u64; 2] {
+        [self.start, self.end]
+    }
+
+    pub fn from_array(arr: [u64; 2]) -> Self {
+        Self::new(arr[0], arr[1])
+    }
+}
```

Comment:
Hmm, did you think about using a built in std::ops::Range or std::ops::InclusiveRange here?

Reply:
I guess since you're not really doing any operations on this on the edge side, it isn't needed.